### DISCUSSION TITLE
Defer: client expectations on subsequent payloads’ path

### DISCUSSION THREAD

**Discussion by BoD** [graphql/defer-stream-wg#46] [Q&A] [answered]
Hi!

Given this schema:

```graphql
type Query {
  product(id: ID!): Product
}

type Product {
  id: ID!
  description: String
}
```

This query:
```graphql
query ProductById {
  product(id: "0") {
    id
    ... on Product @defer {
      description
    }
  }
}
```

could return:

#### Payload 1
```json
{
  "data": {
    "product": {
      "id": "0"
    }
  },
  "hasNext": true
}
```

#### Payload 2
```json
{
  "data": {
    "label": "Product 0"
  },
  "path": [ "product" ],
  "hasNext": false
}
```

But could the response be of this form:

#### Payload 1
same as above

#### Payload 2
```json
{
  "data": {
    "product": {
      "label": "Product 0"
    }
  },
  "hasNext": false
}
```

?

In other words, should subsequent payloads always be rooted at the path where the related `@defer` directive is applied (example 1)? Or is it also possible for them to be rooted above it (example 2)?

**Comment by glasser** [comment, marked as answer]
I'm reasonably sure based on my understanding of the spec that only the first version is correct — but I think the spec should have some more examples/clarity to be explicit that subsequent payloads are always "here's a subtree with the place to splice it in" and not "here's some more tree, rooted at the top, to be merged in".

**Reply by robrichard** [reply]
Yes, this is correct. I'll make sure this is noted in the spec text.

**Reply by robrichard** [reply]
I added the following wording to the response section of the spec

defer: 
> The path segments must point to the location of the result of the field containing the associated `@defer` directive.

stream:
> All but the non-final path segments must
refer to the location of the list field containing the corresponding `@stream`
directive. The final segment of the path list must be a 0-indexed integer. This
integer indicates that this result is set at a range, where the beginning of the
range is at the index of this integer, and the length of the range is the length
of the data.

**Reply by glasser** [reply]
Or to be clear, there actually *is* merging still, but a single level (ie, `data` doesn't replace the entire node at `path`, which would remove the `id` above, but it adds all of its fields there)
